src/controllers/csi/gc/images.go

```diff
@@ -2,78 +2,93 @@ package csigc
+	if err != nil {
+		return err
+	}
+	if len(imagesToDelete) == 0 {
+		log.Info("no shared image dirs to delete on the node")
+		return nil
+	}
 
-	deleteImageCaches(gc.fs, imageCachesToDelete)
+	return deleteImageDirs(gc.fs, imagesToDelete)
 }
 
-func getImageCacheDirs(fs afero.Fs) ([]os.FileInfo, error) {
-	imageDirs, err := afero.Afero{Fs: fs}.ReadDir(image.CacheDir)
+func (gc *CSIGarbageCollector) getSharedImageDirs() ([]os.FileInfo, error) {
+	imageDirs, err := afero.Afero{Fs: gc.fs}.ReadDir(gc.path.AgentSharedBinaryDirBase())
 	if os.IsNotExist(err) {
-		log.Info("No image cache to clean up")
 		return nil, nil
 	}
 	if err != nil {
-		log.Error(err, "Failed to read image cache directory")
-		return nil, err
+		log.Info("failed to read shared image directory")
+		return nil, errors.WithStack(err)
 	}
-	return imageDirs, err
+	return imageDirs, nil
 }
 
-func collectUnusedImageCaches(fs afero.Fs, imageDirs []os.FileInfo) []string {
+func (gc *CSIGarbageCollector) collectUnusedImageDirs(imageDirs []os.FileInfo) ([]string, error) {
 	var toDelete []string
+	usedImageDigests, err := gc.getUsedImageDigests()
+	if err != nil {
+		log.Info("failed to get the used image digests")
+		return nil, err
+	}
 	for _, imageDir := range imageDirs {
 		if !imageDir.IsDir() {
 			continue
 		}
-		modificationTime, err := getRelevantModificationTime(fs, imageDir.Name())
-		if err != nil {
-			log.Error(err, "failed to get modification time of image cache", "imageDir", imageDir.Name())
-			continue
-		}
-		if time.Since(modificationTime) > maxImageAge {
-			toDelete = append(toDelete, filepath.Join(image.CacheDir, imageDir.Name()))
+		imageDigest := imageDir.Name()
+		if !usedImageDigests[imageDigest] {
+			toDelete = append(toDelete, gc.path.AgentSharedBinaryDirForImage(imageDigest))
 		}
 	}
-	return toDelete
+	return toDelete, nil
 }
 
-// getRelevantModificationTime returns the last modification time of an image in the cache
-// an image is not a single file but a directory of several files.
-// Most of which don't change when the image is tried to be pulled again.
-// The index.json of the image is a file that is modified in reliable way.(after every pull)
-func getRelevantModificationTime(fs afero.Fs, imageDir string) (time.Time, error) {
-	var modificationTime time.Time
-	indexPath := filepath.Join(image.CacheDir, imageDir, "index.json")
-	indexFileInfo, err := fs.Stat(indexPath)
+func (gc *CSIGarbageCollector) getUsedImageDigests() (map[string]bool, error) {
+	usedImageDigests, err := gc.db.GetUsedImageDigests()
 	if err != nil {
-		return modificationTime, err
+		log.Info("failed to get the used image digests")
+		return nil, err
+	}
+
+	// the version of a Volume is the imageDigest if the shared image was used during mount
+	// this will still can contain versions that are not imageDigests,
+	// however this shouldn't cause issues as those versions doesn't matter in this context
+	usedVersions, err := gc.db.GetAllUsedVersions()
```

Comment:
if only digests are needed, why isn't `GetUsedImageDigests` used instead?

Reply:
the version in the volume is the digest if the unpacked shared image is mounted, so to not kick the dir out of running user containers if the dynakube(s) was removed